### PR TITLE
impl(common): TimerQueue atomic callbacks

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #10553

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10563)

### CODE REVIEW

File: google/cloud/internal/timer_queue.h [hunk 1]
```diff
@@ -72,10 +75,70 @@ class TimerQueue : public std::enable_shared_from_this<TimerQueue> {
...
+   * - The timer expires, in which case the future is satisfied with a
+   *   successful `StatusOr`.
+   * - The timer queue is shutdown (or was shutdown), in which case the future
+   *   is satisfied with an error status.
+   */
   future<StatusOr<std::chrono::system_clock::time_point>> Schedule(
       std::chrono::system_clock::time_point tp);
 
+  /**
+   * Adds a timer to the queue and atomically attach a callback to the timer.
```
Comment (devbww) [inline google/cloud/internal/timer_queue.h:91]: "atomically" probably deserves a little explanation.  Perhaps just phrase it in terms of who might make the callback?

Comment (coryan) [reply google/cloud/internal/timer_queue.h]: Done.

File: google/cloud/internal/timer_queue.h [hunk 2]
```diff
@@ -72,10 +75,70 @@ class TimerQueue : public std::enable_shared_from_this<TimerQueue> {
...
+   * When the functor is called its future will be already satisfied. The
+   * `StatusOr<>` value in may contain an error.  This can be used to detect
+   * if the timer queue is shutdown.  For example, implementing
+   * periodic timers may be implemented as:
+   *
+   * @code {.cpp}
+   * void StartPeriodicTimer(
+   *     std::shared_ptr<TimerQueue> tq, std::chrono::milliseconds period,
+   *     std::function<void()> action) {
+   *   auto functor = [w = std::weak_ptr<TimerQueue>(tq), period, action](
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue.h]: Done

File: google/cloud/internal/timer_queue.h [hunk 3]
```diff
@@ -72,10 +75,70 @@ class TimerQueue : public std::enable_shared_from_this<TimerQueue> {
...
+   * Adds a timer to the queue and atomically attach a callback to the timer.
+   *
+   * This creates a new timer that expires at @p tp, and atomically attaches
+   * @p functor to be invoked when the timer expires. The functor must be able
+   * to consume `future<StatusOr<std::chrono::system_clock::time_point>>`.
+   *
+   * When the functor is called its future will be already satisfied. The
+   * `StatusOr<>` value in may contain an error.  This can be used to detect
+   * if the timer queue is shutdown.  For example, implementing
+   * periodic timers may be implemented as:
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue.h]: Fixed

File: google/cloud/internal/timer_queue.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue.h]: Fixed

File: google/cloud/internal/timer_queue.h [hunk 4]
```diff
@@ -111,6 +174,23 @@ class TimerQueue : public std::enable_shared_from_this<TimerQueue> {
   // Cancels a timer.
   void Cancel(KeyType key);
 
+  // Creates a satisfied future indicating the queue is shutdown.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue.h]: Ooops, changed the type and not the comments. Fixed.

File: google/cloud/internal/timer_queue.h [hunk 5]
```diff
@@ -72,10 +75,70 @@ class TimerQueue : public std::enable_shared_from_this<TimerQueue> {
   TimerQueue(TimerQueue&&) = delete;
   TimerQueue& operator=(TimerQueue&&) = delete;
 
-  /// Adds a timer to the queue.
+  /**
+   * Adds a timer to the queue.
+   *
+   * The future returned by this function is satisfied when either:
+   * - The timer expires, in which case the future is satisfied with a
+   *   successful `StatusOr`.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue.h]: Done

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/timer_queue.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/timer_queue.h]: Yes, I meant that during development I changed the type and not the comments.  I am happy as-is.

**Pull request merged**